pymc3/backends/arviz.py

```diff
@@ -0,0 +1,606 @@
+            if hasattr(obs, "observations"):
+                aux_obs = obs.observations
+                observations[obs.name] = (
+                    aux_obs.get_value() if hasattr(aux_obs, "get_value") else aux_obs
+                )
+            elif hasattr(obs, "data") and self.density_dist_obs:
+                for key, val in obs.data.items():
+                    aux_obs = val.eval() if hasattr(val, "eval") else val
+                    multi_observations[key] = (
+                        aux_obs.get_value() if hasattr(aux_obs, "get_value") else aux_obs
+                    )
+        return observations, multi_observations
+
+    def split_trace(self) -> Tuple[Union[None, "MultiTrace"], Union[None, "MultiTrace"]]:
+        """Split MultiTrace object into posterior and warmup.
+
+        Returns
+        -------
+        trace_posterior: MultiTrace or None
+            The slice of the trace corresponding to the posterior. If the posterior
+            trace is empty, None is returned
+        trace_warmup: MultiTrace or None
+            The slice of the trace corresponding to the warmup. If the warmup trace is
+            empty or ``save_warmup=False``, None is returned
+        """
+        trace_posterior = None
+        trace_warmup = None
+        if self.save_warmup and self.ntune > 0:
+            trace_warmup = self.trace[: self.ntune]
+        if self.ndraws > 0:
+            trace_posterior = self.trace[self.ntune :]
+        return trace_posterior, trace_warmup
+
+    def log_likelihood_vals_point(self, point, var, log_like_fun):
+        """Compute log likelihood for each observed point."""
+        log_like_val = utils.one_de(log_like_fun(point))
+        if var.missing_values:
+            mask = var.observations.mask
+            if np.ndim(mask) > np.ndim(log_like_val):
+                mask = np.any(mask, axis=-1)
+            log_like_val = np.where(mask, np.nan, log_like_val)
+        return log_like_val
+
+    def _extract_log_likelihood(self, trace):
+        """Compute log likelihood of each observation."""
+        if self.trace is None:
+            return None
+        if self.model is None:
+            return None
+
+        if self.log_likelihood is True:
+            cached = [(var, var.logp_elemwise) for var in self.model.observed_RVs]
+        else:
+            cached = [
+                (var, var.logp_elemwise)
+                for var in self.model.observed_RVs
+                if var.name in self.log_likelihood
+            ]
+        log_likelihood_dict = _DefaultTrace(len(trace.chains))
+        for var, log_like_fun in cached:
+            for chain in trace.chains:
+                log_like_chain = [
+                    self.log_likelihood_vals_point(point, var, log_like_fun)
+                    for point in trace.points([chain])
+                ]
+                log_likelihood_dict.insert(var.name, np.stack(log_like_chain), chain)
+        return log_likelihood_dict.trace_dict
```

Comment:
I am hoping that all this will go away and pointwise log likelihood can now be evaluated in a vectorized way, looping over variables to loop over chains to loop over draws should not be the way to go.

Reply:
Yes, it should be possible to evaluate a likelihood in a vectorized fashion.  You'll still have to compile an Aesara function with a fixed set of dimensions for the input, but, assuming that the number/order of the input dimensions are always the same (e.g. chains + samples + var dims), that need only be done once for a given model.